src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -252,6 +324,83 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
+            peerTwoTip.ChainWork.Should().BeEquivalentTo(chainTip.ChainWork);
+
+            // Peer C / Peer 3 presents an alternative chain with a forkpoint below B's Tip.
+            BlockHeader forkedBlock = testContext.GetNewBlock(chainTip.Previous.Previous).Header;
+            chainedHeaderTree.ConnectNewHeaders(3, new List<BlockHeader>() { forkedBlock });
+            ChainedHeader peerThreeTip = chainedHeaderTree.GetChainedHeaderByPeerId(3);
+
+            ChainedHeader fork = chainTip.FindFork(peerThreeTip);
+            fork.Height.Should().BeLessThan(peerTwoTip.Height);
+            
+            // Peer A / PEER_ONE prolongs C's chain.
+            const int numberOfBlocksToExtend = 7;
+            ChainedHeader peerOneTip = testContext.ExtendAChain(numberOfBlocksToExtend, peerThreeTip);
+            List < BlockHeader > listOfPeerOnesHeaders = testContext.ChainedHeaderToList(peerOneTip, numberOfBlocksToExtend + 2 /* (forked 2 back) */);
+            
+            // Sends all headers from the fork point
+            fork.HashBlock.Should().BeEquivalentTo(listOfPeerOnesHeaders[0].GetHash());
+
+            // With few valid and one invalid.
+            foreach (var header in listOfPeerOnesHeaders.GetRange(3, 6))
+            {
+                testContext.ChainStateMock.Setup(x => x.MarkBlockInvalid(header.GetHash(), null));
+            }
+
+            chainedHeaderTree.ConnectNewHeaders(1, listOfPeerOnesHeaders);
+
+            // Peer A / PEER_ONE's fork should now have the longest chain
+            peerOneTip.Height.Should().BeGreaterThan(peerThreeTip.Height);
+        }
+
+        /// <summary>
+        /// Issue 9 @ Create chained header tree component #1321
+        /// Check that everything is always consumed.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_MultiplePeers_CheckEverythingIsConsumed()
+        {
+            var testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+            ChainedHeader chainTip = testContext.ExtendAChain(11);
+            chainedHeaderTree.Initialize(chainTip, true);
+            List < BlockHeader > listOfNewHeaders = testContext.ChainedHeaderToList(chainTip, 11);
+
+            List<BlockHeader> peer1Headers = listOfNewHeaders.GetRange(0, 6);
+            ConnectNewHeadersResult connectedNewHeadersPeer1 = chainedHeaderTree.ConnectNewHeaders(1, peer1Headers);
+
+            // 5 unprocessed headers
+            Assert.True(chainTip.Height - connectedNewHeadersPeer1.Consumed.Height == 5);
+
+            List<BlockHeader> peer2Headers = listOfNewHeaders.GetRange(6, 5);
+            ConnectNewHeadersResult connectedNewHeadersPeer2 = chainedHeaderTree.ConnectNewHeaders(2, peer2Headers);
+
+            // All headers consumed
+            Assert.True(connectedNewHeadersPeer2.Consumed.Height == chainTip.Height);
+        }
```

Comment:
So currently we have two scenarios here, first one is that we submit a first part of the chain and we check that it is consumed. Then we add something more to this chain and check that it is consumed.

Now we should add couple more scenarios:

– We should now submit such a list of headers that presents some of those that are already in the tree as well as some that are not in the tree yet.
– Then we should submit list of headers that present some of those that are already in the tree and then the rest of the headers form the new fork the tree.
– Then we should submit list of headers in which nothing is already in the tree but it forms a fork
– Finally, we should submit headers that are all already in the tree.

Reply:
All these scenarios now present